js/private/js_binary.bzl

```diff
@@ -123,15 +122,29 @@ _ATTRS = {
         doc = """Environment variables of the action.
 
         Subject to [$(location)](https://bazel.build/reference/be/make-variables#predefined_label_variables)
-        and ["Make variable"](https://bazel.build/reference/be/make-variables) substitution.
+        and ["Make variable"](https://bazel.build/reference/be/make-variables) substitution if `expand_env` is set to True.
         """,
     ),
+    "expand_args": attr.bool(
+        default = True,
+        doc = """Enables [$(location)](https://bazel.build/reference/be/make-variables#predefined_label_variables)
+        and ["Make variable"](https://bazel.build/reference/be/make-variables) substitution for `fixed_args`.
+
+        This comes at some analysis-time cost.""",
+    ),
+    "expand_env": attr.bool(
+        default = True,
+        doc = """Enables [$(location)](https://bazel.build/reference/be/make-variables#predefined_label_variables)
+        and ["Make variable"](https://bazel.build/reference/be/make-variables) substitution for `env`.
+
+        This comes at some analysis-time cost.""",
```

Comment:
nit: This comes at some analysis-time cost even for a set of envs that does not have any expansions.